flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/ChangelogStorageMetricsTest.java

```
@@ -295,6 +346,55 @@ public void close() {
         }
     }
 
+    private static class WaitingMaxAttemptUploader implements StateChangeUploader {
+        private final ConcurrentHashMap<UploadTask, CountDownLatch> remainingAttemptsPerTask;
+        private final int maxAttempts;
+
+        public WaitingMaxAttemptUploader(int maxAttempts) {
+            if (maxAttempts < 1) {
+                throw new IllegalArgumentException("maxAttempts < 0");
+            }
+            this.maxAttempts = maxAttempts;
+            this.remainingAttemptsPerTask = new ConcurrentHashMap<>();
+        }
+
+        @Override
+        public UploadTasksResult upload(Collection<UploadTask> tasks) throws IOException {
+
+            for (UploadTask uploadTask : tasks) {
+                CountDownLatch remainingAttempts = remainingAttemptsPerTask.get(uploadTask);
+                if (remainingAttempts == null) {
+                    remainingAttempts = new CountDownLatch(maxAttempts - 1);
+                    remainingAttemptsPerTask.put(uploadTask, remainingAttempts);
+                } else {
+                    remainingAttempts.countDown();
+                }
+            }
+            for (UploadTask uploadTask : tasks) {
+                CountDownLatch remainingAttempts = remainingAttemptsPerTask.get(uploadTask);
+                try {
+                    remainingAttempts.await();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
```

Should we throw an exception here (in addition to thread.interrupt)?